### PR TITLE
Add support for math integers

### LINKED ISSUE #607: Nat/Int library
We want:
 - a type `Int` on which we can do arithmetic from Rust;
 - this type should be a newtype around i128 or bigint;
 - the translation should map to F*'s `int` (or coq `int`), and all operations on `Int` should be mapped to native ones in F*
 - we need to be able to lift machine integers to ints and lift down

### PR DESCRIPTION
**Pull request by W95Psp**
This PR adds support for math integers. Will fix #607.

This PR:
 - introduces a `Int` type  (for math integers) that implements common operations;
 - translate `Int` as primitive `int` in F*, and same thing for methods on `Int`;
 - introduces a Specialize phase that "monomorphize" operations (e.g. std's add becomes a specific function);
 - provides a `int` macro for math int literals;
 - adds tests for math ints and their operations.

### CODE REVIEW

File: hax-lib/Cargo.toml [hunk 1]
```diff
@@ -12,8 +12,6 @@ description = "Hax-specific helpers for Rust programs"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-hax-lib-macros = { path = "../hax-lib-macros", optional = true }
-
-[features]
```
Comment (franziskuskiefer) [inline hax-lib/Cargo.toml:17]: Where did the features go?

File: hax-lib/Cargo.toml [hunk 2]
```diff
@@ -12,8 +12,6 @@ description = "Hax-specific helpers for Rust programs"
...
-default = ["macros"]
-macros = ["dep:hax-lib-macros"]
+hax-lib-macros = { path = "../hax-lib-macros" }
```
Comment (franziskuskiefer) [inline hax-lib/Cargo.toml:15]: This needs a version.

File: hax-lib/Cargo.toml [hunk 3]
```diff
@@ -12,8 +12,6 @@ description = "Hax-specific helpers for Rust programs"
...
+num-bigint = { version = "0.4.3", default-features = false, features = [] }
```
Comment (franziskuskiefer) [inline hax-lib/Cargo.toml:16]: Using both, `default-features` and `features` seems unnecessary.

File: hax-lib-macros/src/lib.rs [hunk 4]
```diff
@@ -541,6 +541,35 @@ pub fn pv_handwritten(_attr: pm::TokenStream, item: pm::TokenStream) -> pm::Toke
     quote! {#attr #item}.into()
 }
 
+/// Create a mathematical integer. This macro expects a integer
+/// literal that consists in an optional minus sign followed by one or
+/// more digits.
+#[proc_macro_error]
+#[proc_macro]
+pub fn int(payload: pm::TokenStream) -> pm::TokenStream {
```
Comment (franziskuskiefer) [inline hax-lib-macros/src/lib.rs:549]: I don't see a `hax` config. Where is the code compiled out for regular builds?

File: hax-lib/src/int/bigint.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+//! `BYTES`.
+//! Its interface provides bridges to `num_bigint::BigInt`.
+
+#[crate::exclude]
+mod no_extraction {
+    const BYTES: usize = 128;
+    #[derive(Copy, Clone)]
+    pub struct BigInt {
+        sign: num_bigint::Sign,
+        data: [u8; BYTES],
```
Comment (franziskuskiefer) [inline hax-lib/src/int/bigint.rs:12]: 1024 bits seems a little low. This should be at least 1024 bytes.

File: hax-lib/src/int/bigint.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    }
```
Comment (franziskuskiefer) [inline hax-lib/src/int/bigint.rs:13]:
```suggestion
    }

```

File: hax-lib/src/int/bigint.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    impl BigInt {
+        pub fn new(i: &num_bigint::BigInt) -> Self {
```
Comment (franziskuskiefer) [inline hax-lib/src/int/bigint.rs:15]: All public functions need doc comments.

File: hax-lib/src/int/bigint.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            let (sign, data) = i.to_bytes_be();
+            let data = data.try_into().expect(
+                "`copiable_bigint::BigInt::new`: too big, please consider increasing `BYTES`",
+            );
+            BigInt { sign, data }
+        }
+        pub fn get(self) -> num_bigint::BigInt {
```
Comment (franziskuskiefer) [inline hax-lib/src/int/bigint.rs:22]:
```suggestion

        pub fn get(self) -> num_bigint::BigInt {
```

File: hax-lib/src/int/mod.rs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+impl Int {
+    /// Raises `2` at the power `self`
+    pub fn pow2(self) -> Self {
+        let exponent = self.get().to_u32().expect("Exponent doesn't fit in a u32");
+        Self::new(num_bigint::BigInt::from(2u8).pow(exponent))
+    }
+
+    /// Constructs a `Int` out of a string literal. This function
+    /// assumes its argument consists only of digits, with optionally
```
Comment (franziskuskiefer) [inline hax-lib/src/int/mod.rs:64]: What type of digits? Hex?

File: hax-lib/src/int/mod.rs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    type Output = Self;
+
+    fn div(self, other: Self) -> Self::Output {
+        Self::new(self.get() / other.get())
+    }
+}
+
+impl Int {
+    /// Raises `2` at the power `self`
+    pub fn pow2(self) -> Self {
```
Comment (franziskuskiefer) [inline hax-lib/src/int/mod.rs:58]: Do we need this function anywhere?

File: hax-lib/src/lib.rs [hunk 11]
```diff
@@ -15,6 +15,8 @@
 
 #![no_std]
 
+pub mod int;
```
Comment (franziskuskiefer) [inline hax-lib/src/lib.rs:18]: How is this compiled out when not in `cfg(hax)`?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, a couple notes.

File: hax-lib/src/int/mod.rs (hunk 10 above)
Comment (W95Psp) [reply hax-lib/src/int/mod.rs:58]: Yes, we use that a lot for Kyber and its spec

File: hax-lib/Cargo.toml (hunk 3 above)
Comment (W95Psp) [reply hax-lib/Cargo.toml:16]: yeah, sorry, was fighting with CI, fixing that

File: hax-lib/src/lib.rs (hunk 11 above)
Comment (W95Psp) [reply hax-lib/src/lib.rs:18]: Postponing that to another PR, #639

File: hax-lib-macros/src/lib.rs (hunk 4 above)
Comment (W95Psp) [reply hax-lib-macros/src/lib.rs:549]: Postponing that to another PR, #639

**Review by franziskuskiefer — APPROVED** [review]

File: engine/lib/phases/phase_specialize.ml [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+          (* let is_int = is_type0 Hax_lib__int__Int *)
+          (* let is_ref_int = is_ref_type0 Hax_lib__int__Int *)
+
+          (* let mk_ref_int n f f' = *)
+          (*   { *)
+          (*     fn = f; *)
+          (*     fn_replace = f'; *)
+          (*     args = List.init n ~f:(fun _ -> is_ref_int); *)
+          (*   } *)
```
[the remark this answers is not in the document]
Comment (W95Psp) [reply engine/lib/phases/phase_specialize.ml]: Oh, thanks, I re-wrote the logic of this pattern thing, forgot to clean up!
I cleaned up and added some documentation there.

**Review by karthikbhargavan — APPROVED** [review]

**Pull request merged**